toolkit/tools/imagegen/installutils/installutils.go

```diff
@@ -1693,29 +1695,56 @@ func selinuxRelabelFiles(systemConfig configuration.SystemConfig, installChroot
 	selinuxType := strings.TrimSpace(stdout)
 	fileContextPath := fmt.Sprintf(fileContextBasePath, selinuxType)
 
-	logger.Log.Debugf("Running setfiles to apply SELinux labels on mount points: %v", listOfMountsToLabel)
-	err = installChroot.UnsafeRun(func() error {
-		args := []string{"-m", "-v", fileContextPath}
-		args = append(args, listOfMountsToLabel...)
-
-		// We only want to print basic info, filter out the real output unless at trace level (Execute call handles that)
-		files := 0
-		lastFile := ""
-		onStdout := func(args ...interface{}) {
-			if len(args) > 0 {
-				files++
-				lastFile = fmt.Sprintf("%v", args)
+	targetRootPath := "/mnt/_bindmountroot"
+
+	for _, mountToLabel := range listOfMountsToLabel {
+		logger.Log.Debugf("Running setfiles to apply SELinux labels on mount points: %v", mountToLabel)
+
+		// Create a bind mount so that the filesystem's files can be labelled without interference from other mounts.
+		sourceFullPath := filepath.Join(installChroot.RootDir(), mountToLabel)
+		targetPath := filepath.Join(targetRootPath, mountToLabel)
+		targetFullPath := filepath.Join(installChroot.RootDir(), targetPath)
+
+		bindMount, err := safemount.NewMount(sourceFullPath, targetFullPath, "", unix.MS_BIND, "", true)
+		if err != nil {
+			return fmt.Errorf("failed to bind mount (%s) while SELinux labeling:\n%w", mountToLabel, err)
+		}
+		defer bindMount.Close()
+
+		err = installChroot.UnsafeRun(func() error {
+			// We only want to print basic info, filter out the real output unless at trace level (Execute call handles that)
+			files := 0
+			lastFile := ""
+			onStdout := func(args ...interface{}) {
+				if len(args) > 0 {
+					files++
+					lastFile = fmt.Sprintf("%v", args)
+				}
+				if (files % 1000) == 0 {
+					ReportActionf("SELinux: labelled %d files", files)
+				}
 			}
-			if (files % 1000) == 0 {
-				ReportActionf("SELinux: labelled %d files", files)
+			err := shell.ExecuteLiveWithCallback(onStdout, logger.Log.Warn, squashErrors, "setfiles", "-m", "-v", "-r",
+				targetRootPath, fileContextPath, targetPath)
+			if err != nil {
+				return fmt.Errorf("failed while labeling files (last file: %s) %w", lastFile, err)
 			}
+			return err
+		})
+		if err != nil {
+			return err
 		}
-		err := shell.ExecuteLiveWithCallback(onStdout, logger.Log.Warn, squashErrors, "setfiles", args...)
+
+		err = bindMount.CleanClose()
 		if err != nil {
-			return fmt.Errorf("failed while labeling files (last file: %s) %w", lastFile, err)
+			return err
 		}
-		return err
-	})
+	}
+
+	err = os.RemoveAll(targetRootPath)
```

Comment:
I'm 50/50 on this, but part of me things this could be a deferred function... but on the other hand if this fails the build is dead anyways so... 🤷

Reply:
Considering that this is within the image and not the host, I think it is fine to leave it as is. Also, using `defer` has a tendency to encourage people to ignore errors.